### PR TITLE
Uniffi simple message relay

### PR DESCRIPTION
**Pull request by swaptr**

### CODE REVIEW

File: uniffi_aries_vcx/core/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ path = "uniffi-bindgen.rs"
 uniffi = "0.23.0"
 aries-vcx = { path = "../../aries_vcx" }
 tokio = { version = "1.24.1" }
+libvdrtools = { path = "../../libvdrtools", features=["default"] }
```
Comment (gmulhearn) [inline uniffi_aries_vcx/core/Cargo.toml:22]: why is this needed??

File: uniffi_aries_vcx/core/src/core/unpack_message.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        let wallet = profile_holder.inner.inject_wallet();
+        let unpacked_bytes = wallet.unpack_message(&packed_bytes).await?;
+        let unpacked_string = match String::from_utf8(unpacked_bytes) {
+            Ok(str) => str,
+            Err(err) => todo!(),
+        };
+        let value = match serde_json::from_str::<UnpackMessage>(&unpacked_string) {
+            Ok(str) => str,
+            Err(err) => todo!(),
+        };
```
Comment (gmulhearn) [inline uniffi_aries_vcx/core/src/core/unpack_message.rs:31]: can just escape `serde` errors with `?`. as we have already defined a serde error -> VcxUniFFI error transformer

File: uniffi_aries_vcx/core/src/core/unpack_message.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+use std::sync::Arc;
+
+use aries_vcx::errors::error::{AriesVcxError, AriesVcxErrorKind};
+use vdrtools::domain::crypto::pack::UnpackMessage;
```
Comment (gmulhearn) [inline uniffi_aries_vcx/core/src/core/unpack_message.rs:4]: preferably create your own struct for this rather than importing from libvdrtools. we should try to avoid direct libvdrtools dependencies

Comment (gmulhearn) [reply uniffi_aries_vcx/core/src/core/unpack_message.rs:4]: you could just copy paste libvdrtool's definition for the struct;
```
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct UnpackMessage {
    pub message: String,
    pub recipient_verkey: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sender_verkey: Option<String>,
}
```

File: uniffi_aries_vcx/core/src/core/unpack_message.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+use super::profile::ProfileHolder;
+use crate::{errors::error::VcxUniFFIResult, runtime::block_on};
+
+pub fn unpack_message(profile_holder: Arc<ProfileHolder>, packed_msg: Vec<u8>) -> VcxUniFFIResult<Vec<u8>> {
```
Comment (gmulhearn) [inline uniffi_aries_vcx/core/src/core/unpack_message.rs:9]: if you wanted, could do it this way:
```
pub fn unpack_message(profile_holder: Arc<ProfileHolder>, packed_msg: String) -> VcxUniFFIResult<UnpackMessage> {
...
}
```

in kotlin will look like:
```
data class Unpackmessage(message: String, recip: String, sender: Option<String>)
```

File: uniffi_aries_vcx/core/Cargo.toml (hunk 1 above)
Comment (swaptr) [reply uniffi_aries_vcx/core/Cargo.toml]: Was used for accessing the `UnpackMessage` trait for deserialization. Now fixed with an implementation of it inside the uniffi crate.

File: uniffi_aries_vcx/demo/app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -1,3 +1,4 @@
 <resources>
     <string name="app_name">demo</string>
-</resources>
\ No newline at end of file
+    <string name="serviceEndpoint">https://b199-27-57-116-96.ngrok-free.app</string>
```
Comment (swaptr) [inline uniffi_aries_vcx/demo/app/src/main/res/values/strings.xml:3]: Does this need to be set as an env var? Or any other secure and flexible way?

Comment (gmulhearn) [reply uniffi_aries_vcx/demo/app/src/main/res/values/strings.xml:3]: to be honest, i'd just make a `Constants.kt` file, and put this value in as a `BASE_RELAY_ENDPOINT` const val, and add a comment above it explaining what it means

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt [hunk 6]
```diff
@@ -66,9 +74,13 @@ fun ScanScreen(connection: Connection, profileHolder: ProfileHolder) {
                         )
                         connection.sendRequest(
                             profileHolder,
-                            "https://google.com",
+                            "$BASE_RELAY_ENDPOINT/send_user_message/${walletConfig.walletKey}",
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:77]: using walletKey as the unique ID is very suspicious; wallet key is meant to be a unique password for your device wallet. I'd recommend we just use something else hardcoded as the ID. e.g. `demo-user-1`

Comment (swaptr) [reply uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:77]: Oh no I dont know what I was thinking, This is supposed to be a user id, right?

**Review by gmulhearn — DISMISSED** [review]
Nice work. This will be good to merge after we stop using `walletKey` as the user id; for now we can just use a hardcoded string: `demo-user-1` for example.

And also as discussed, in a later PR we will look into cleaning up the app logic by putting a lot of the logic into a global "controller" class. this will let us clean up some of the HTTP polling logic currently within the HomeScreen

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt [hunk 7]
```diff
@@ -1,37 +1,76 @@
...
-        storageCredentials = null,
-        rekey = null,
-        rekeyDerivationMethod = null
-    )
+    var flagKeepFetching by remember {
+        mutableStateOf(true)
+    }
+
+    val request = Request.Builder()
+        .url("$BASE_RELAY_ENDPOINT/pop_user_message/${walletConfig.walletKey}")
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt:49]: still an issue here with walletKey being used. should match the other string. you could also make it into some constant string;

`const val RELAY_USER_ID = "demo-user-1"

**Review by gmulhearn — APPROVED** [review]
